File: bench/algorithm/regex-redux/3-m.ml
(* The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * regex-dna program contributed by Christophe TROESTLER
 * updated by Roman Kashitsyn: use Bytes instead of String
 * updated by Gaëtan Dubreil: use the Re library and parallelize processing
 *)

 open Printf

 let variants =
   [
     "agggtaaa|tttaccct";
     "[cgt]gggtaaa|tttaccc[acg]";
     "a[act]ggtaaa|tttacc[agt]t";
     "ag[act]gtaaa|tttac[agt]ct";
     "agg[act]taaa|ttta[agt]cct";
     "aggg[acg]aaa|ttt[cgt]ccct";
     "agggt[cgt]aa|tt[acg]accct";
     "agggta[cgt]a|t[acg]taccct";
     "agggtaa[cgt]|[acg]ttaccct";
   ]
 
 let subst =
   [
     ("tHa[Nt]", "<4>");
     ("aND|caN|Ha[DS]|WaS", "<3>");
     ("a[NSt]|BY", "<2>");
     ("<[^>]*>", "|");
     ("\\|[^|][^|]*\\|", "-");
   ]
 
 (* Read all of a redirected FASTA format file from file. *)
 let file_data, file_length =
   let n_args = Array.length Sys.argv in
   let file_name =
     match n_args > 1 with true -> Array.get Sys.argv 1 | false -> "25000_in"
   in
   let file_channel = open_in file_name in
   let b = Buffer.create 0xFFFF in
   let s = Bytes.create 0xFFF in
   let r = ref 1 in
   while !r > 0 do
     r := input file_channel s 0 0xFFF;
     Buffer.add_substring b (Bytes.unsafe_to_string s) 0 !r
   done;
   (Buffer.contents b, Buffer.length b)
 
 (* Remove FASTA sequence descriptions and all linefeed characters.  *)
 let dna = Re.replace_string (Re.Pcre.regexp ">.*\n|\n") ~by:"" file_data
 
 let code_length = String.length dna
 
 (* Count matches of [re]. *)
 let count re s =
   let re = Re.Pcre.regexp re in
   let i = ref 0 in
   let n = ref 0 in
   try
     while true do
       let grps = Re.exec ~pos:!i re s in
       i := Re.Group.stop grps 0;
       incr n
     done;
     assert false
   with Not_found -> !n
 
 let () =
   if Unix.fork () = 0 then
     List.iter (fun re -> printf "%s %i\n" re (count re dna)) variants
   else
     let b = ref dna in
     List.iter
       (fun (re, s) -> b := Re.replace_string (Re.Pcre.regexp re) ~by:s !b)
       subst;
 
     ignore (Unix.wait ());
     printf "\n%i\n%i\n%i\n" file_length code_length (String.length !b)
 

File: bench/algorithm/regex-redux/3-m.rb
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
#
# Rewrite for regex-redux by Aaron Tavistock
# array not dictionary by Isaac Gouy

class RegexRedux

    MATCHERS = [
      /agggtaaa|tttaccct/,
      /[cgt]gggtaaa|tttaccc[acg]/,
      /a[act]ggtaaa|tttacc[agt]t/,
      /ag[act]gtaaa|tttac[agt]ct/,
      /agg[act]taaa|ttta[agt]cct/,
      /aggg[acg]aaa|ttt[cgt]ccct/,
      /agggt[cgt]aa|tt[acg]accct/,
      /agggta[cgt]a|t[acg]taccct/,
      /agggtaa[cgt]|[acg]ttaccct/
    ]
  
  # ruby 1.8.7: to iterate in-order use array not dictionary
    FINAL_TRANSFORM = [
      [ /tHa[Nt]/, '<4>' ], 
      [ /aND|caN|Ha[DS]|WaS/, '<3>' ], 
      [ /a[NSt]|BY/, '<2>' ], 
      [ /<[^>]*>/, '|' ],
      [ /\|[^|][^|]*\|/, '-' ]
    ]
  
    def initialize(seq)
      @seq = seq
      @original_size = @seq.size
      @clean_size = remove_breaks!
      @match_results = match_results
      @final_size = final_transform!
    end
  
    def to_s
      "%s\n\n%d\n%d\n%d" % [
        @match_results.join("\n"),
        @original_size,
        @clean_size,
        @final_size
      ]
    end
  
    def pattern_count(regex)
      count = 0
      @seq.scan(regex) { count += 1 }
      "#{regex.source} #{count}"
    end 
  
    def forked_pattern_count(regex)
      reader, writer = IO.pipe
      Process.fork do
        reader.close
        writer.write(original_pattern_count(regex))
      end
  
      writer.close
      results = reader.read
      reader.close
    
      results
    end
  
  if (RUBY_PLATFORM != 'java') 
      alias_method :original_pattern_count, :pattern_count
      alias_method :pattern_count, :forked_pattern_count
    end
  
    def remove_breaks!
      @seq.gsub!(/>.*\n|\n/, '')
      @seq.size
    end
  
    def match_results
      threads = MATCHERS.map do |matcher|
        Thread.new do
          Thread.current[:result] = pattern_count(matcher)
        end
      end
      threads.each(&:join)
      threads.map { |t| t[:result] }
    end
  
    def final_transform!
      FINAL_TRANSFORM.each { |f,r| @seq.gsub!(f,r) }
      @seq.size
    end
  
  end
  
  fileName = ARGV[0] || "25000_in"
  regex_redux = RegexRedux.new(IO.readlines(fileName).join)
  puts regex_redux.to_s
